cmd/server/main.go

```diff
@@ -0,0 +1,215 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/netip"
+	"os"
+	"sync"
+	"sync/atomic"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	pb "github.com/CobaltCause/go-remote/v2/gen/proto"
+	pm "github.com/CobaltCause/go-remote/v2/pkg"
+	"github.com/alecthomas/kong"
+)
+
+type grpcServer struct {
+	pb.UnimplementedGoRemoteServer
+
+	processManager pm.ProcessManager
+}
+
+func (s *grpcServer) Start(_ context.Context, req *pb.StartRequest) (
+	*pb.StartResponse, error,
+) {
+	// Convert [][]byte to []string.
+	args := make([]string, len(req.GetArgs()))
+	for i, arg := range req.GetArgs() {
+		args[i] = string(arg)
+	}
+
+	id, err := s.processManager.Start(string(req.GetPath()), args...)
+
+	if err != nil {
+		log.Print("failed to start process: ", err)
+
+		// Mostly likely the error has to do with the specific value the client
+		// supplied, e.g. a program not being found. Ideally we'd also detect
+		// other error conditions and use an appopriate status code.
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	log.Print("started process with ID: ", id)
+
+	id_uint32 := uint32(id)
+
+	return &pb.StartResponse{
+			Id: &id_uint32,
+		},
+		nil
+}
+
+func (s *grpcServer) Stop(_ context.Context, req *pb.StopRequest) (
+	*emptypb.Empty, error,
+) {
+	err := s.processManager.Stop(int(req.GetId()))
+
+	if err != nil {
+		log.Print("failed to stop process: ", err)
+
+		if errors.Is(err, pm.ErrUnknownID) {
+			return nil, status.Error(codes.NotFound, err.Error())
+		} else {
+			return nil, err
+		}
+	}
+
+	log.Print("stopped process ID: ", req.GetId())
+
+	return &emptypb.Empty{}, nil
+}
+
+func (s *grpcServer) Status(_ context.Context, req *pb.StatusRequest) (
+	*pb.StatusResponse, error,
+) {
+
+	st, err := s.processManager.Status(int(req.GetId()))
+
+	if err != nil {
+		log.Print("failed to get process status: ", err)
+		return nil, status.Error(codes.NotFound, err.Error())
+	}
+
+	log.Print("responding with state of process ID: ", req.GetId())
+
+	switch st.State {
+	case pm.EXITED:
+		exitCode := int32(st.ExitCode)
+		return &pb.StatusResponse{
+				ExitCode: &exitCode,
+				State:    pb.State_EXITED.Enum(),
+			},
+			nil
+	case pm.RUNNING:
+		return &pb.StatusResponse{
+				State: pb.State_RUNNING.Enum(),
+			},
+			nil
+	case pm.STOPPED:
+		return &pb.StatusResponse{
+				State: pb.State_STOPPED.Enum(),
+			},
+			nil
+	default:
+		panic("unreachable")
+	}
+}
+
+func (s *grpcServer) Stream(
+	req *pb.StatusRequest,
+	stream grpc.ServerStreamingServer[pb.Output],
+) error {
+	stdout, stderr := s.processManager.Stream(int(req.GetId()))
```

Comment:
I'm now noticing an unfortunate UX consequence of the design choice to make stdout and stderr separate streams so that they could each be represented by `io.Reader`/`io.ReadCloser`, which is that ordering of messages between stdout and stderr as emitted from the original process is not preserved. This can result in intermingled logs while streaming live output, and also each stream being sent to the client in large chunks while catching up with the history.

To fix this I'd want to change the design from stdout and stderr being separate objects with a `[]byte` inside to something like `[]struct{[]byte, StdoutOrStderr}` (pseudocode), where the inner `[]byte` would contain all data since the last time `write(2)` was called on the stream.

Reply:
Ha, for my taste at least, the somewhat oddly ordered output is fine, and implementing buffers this way tends to lead to designs that avoid other worse problems, e.g. slow reader problems.